lib/eco/reducer.py

```diff
@@ -0,0 +1,153 @@
+from grammars.grammars import lang_dict
+from treemanager import TreeManager
+from utils import KEY_UP as UP, KEY_DOWN as DOWN, KEY_LEFT as LEFT, KEY_RIGHT as RIGHT
+from optparse import OptionParser
+import logging
+import sys, os
+
+class Reducer(object):
+
+    def __init__(self, lang, program, test):
+        grm = lang_dict[language]
+        parser, lexer = grm.load()
+        parser.init_ast()
+        self.parser = parser
+        self.lexer = lexer
+        self.grm = grm
+        self.treemanager = TreeManager()
+        self.treemanager.add_parser(parser, lexer, grm.name)
+        self.treemanager.version = 1
+        self.treemanager.last_saved_version = 1
+
+        self.program = program
+        self.test = test
+        self.pos = 0
+        self.lastexception = None
+
+    def reset(self):
+        self.parser.reset()
+        self.treemanager = TreeManager()
+        self.treemanager.add_parser(self.parser, self.lexer, self.grm)
+
+    def auto(self):
+        logging.info("Running auto-mode")
+        # try to reduce snapshots first
+        while True:
+            self.original = self.test[:] # remember original tests
+            self.reduce("snapshot")
+            if self.test == self.original:
+                logging.info("No more snapshot reductions possible.")
+                break
+        # try single reduce next
+        while True:
+            self.original = self.test[:] # remember original tests
+            self.reduce("single")
+            if self.test == self.original:
+                logging.info("No more single reductions possible.")
+                break
+
+    def reduce(self, mode):
+        self.pos = 0
+        logging.info("Pos: {} Len: {} Start {}-reduce".format(self.pos, len(self.test), mode))
+        try:
+            # Run without changes to store exception
+            self.run()
+        except Exception as e:
+            self.lastexception = e
+        while self.pos < len(self.test):
+            self.delete(mode)
+            try:
+                self.run() # check if first run causes error
+                logging.info("Pos: {} Len: {} Fixed".format(self.pos, len(self.test)))
+                self.revert()
+            except Exception as e:
+                if not self.same_exception(e):
+                    logging.info("Pos: {} Len: {} Wrong exception: '{}'".format(self.pos, len(self.test), e))
+                    self.revert()
+                    continue
+                else:
+                    logging.info("Pos: {} Len: {} Still failing: '{}'".format(self.pos, len(self.test), e))
+                    self.save() # save progress
+        self.save()
+
+    def same_exception(self, e):
+        if not self.lastexception:
+            self.lastexception = e
+            return True
+        if type(self.lastexception) is type(e) and \
+                self.lastexception.args == e.args:
```

Comment:
Vim's autoindenter is sometimes broken in my experience :( Best to fix this one by hand.